### ISSUE TITLE
Dealing with "Satellite Time"

### ISSUE THREAD

**Issue by johnnyman727** [tessel/gps-a2235h#3]
We currently send back a date and time like the following:

```
Current GPS date and time:
  Year:      2014
  Month:     05
  Day:       25
  Timestamp:     21:01:36.000
```

That timestamp says that it's 9:01pm when it's actually 2:01 pm. 

Is that time parsed wrong? If not, can we get the timezone information? It would be ideal if the time was a `Date` object

**Comment by mattsoftware** [comment]
How do I get the date/time format from the module? At the moment I am only getting timestamp in the form hhmmss.xxx. I would like to get the date out of it as well but so far have been unable to work out how to get it!.

**Comment by jiahuang** [comment]
@mattsoftware,

So most of the GPS processing is actually done in firmware. It's been a while since I did the parsing, but IIRC I only parsed 2 GPS message types: [GPRMC and GPGGA](https://github.com/tessel/firmware/blob/master/src/module_shims/gps-nmea.c#L66). These gps functions are called from lua [here](https://github.com/tessel/firmware/blob/master/src/hw/l_hw.c#L712) and [here](https://github.com/tessel/firmware/blob/master/src/hw/l_hw.c#L1077). 

Then the GPS js code is able to call the functions by binding to the `hw` lib, like in [here](https://github.com/tessel/gps-a2235h/blob/master/index.js#L62).

It looks like you want the [GPZDA](http://aprs.gids.nl/nmea/#zda) parsed as well. In order to do that you'll have to modify the gps-nema shim and add in some additional parsing [here](https://github.com/tessel/firmware/blob/master/src/module_shims/gps-nmea.c#L170).

Then probably add in an additional "get date" function that can be exposed from within JS.

**Comment by mattsoftware** [comment]
Looks like the get date function is already there, but simply calls the get timestamp. I wanted to avoid firmware development, but I think I'll jump in and have a look at this. Thanks for your thorough reply.

**Comment by jiahuang** [comment]
It's getting parsed from the [GPGGA](http://aprs.gids.nl/nmea/#gga) data which doesn't have the full date, only the timestamp.

**Comment by srl295** [comment]
@jiahuang  yeah https://github.com/tessel/t1-firmware/blob/master/src/module_shims/gps-nmea.c#L163 could parse [$GPZDA](http://aprs.gids.nl/nmea/#zda) to get the date, however the code seems to parse [$GPRMC field 9](http://aprs.gids.nl/nmea/#rmc) which should be a "UT date".

**Comment by srl295** [comment]
@jiahuang with the bug above fixed, `gps_get_date()` does seem to return the full date, so `$GPRMC` is coming in. Probably [here](https://github.com/tessel/gps-a2235h/blob/master/index.js#L140) you could add `datestamp: hw.gps_get_date()`
